src/hotspot/share/utilities/vmError.cpp

```diff
@@ -1312,6 +1313,38 @@ int VMError::prepare_log_file(const char* pattern, const char* default_pattern,
   return fd;
 }
 
+// Attempt to set the state to Native in VMError.
+// By setting it to Native, it allows os::fork_and_exec to execute cmd such as jcmd %p.
+// Otherwise, it may end up with a deadlock when cmd tries to synchronize all Java threads
+// at safepoints.
+class VMErrorThreadToNativeFromVM : public StackObj {
+ private:
+   JavaThread* _thread;
+
+ public:
+  VMErrorThreadToNativeFromVM(Thread* t) : _thread(nullptr) {
+    if (t != nullptr && t->is_Java_thread() &&
+        !Threads_lock->owned_by_self()) { // VMError::controlled_crash() grabs Threads_lock.
+      _thread = JavaThread::cast(t);
+      assert(_thread == Thread::current(), "must be current thread");
+      assert(_thread->thread_state() == _thread_in_vm, "must be in VM");
+    }
+
+    if (_thread != nullptr) {
+      assert(!_thread->owns_locks(), "must release all locks when leaving VM");
+      ThreadStateTransition::transition_from_vm(_thread, _thread_in_native);
+    }
+  }
+
+  ~VMErrorThreadToNativeFromVM() {
+    if (_thread != nullptr) {
+      ThreadStateTransition::transition_from_native(_thread, _thread_in_vm);
```

Comment:
Just some thought on this part. Ideally we should avoid calling process_if_requested_with_exit_check() since attempting to process handshakes/stackwatermarks at this point might lead to all sorts of other issues. An alternative could be to just set the original state back and continue. But maybe we don't care about this because we are almost done with the error reporting and the OnError commands were already executed. That last part would argue to move the wrapper before the while loop.

Reply:
I agree with @pchilano (if I understand him correctly) in that I think an RAII object here is not even needed. There is no need to re-instate the prior thread state.

Reply:
As long as we don't call process_if_requested_with_exit_check() I think either way should be fine, i.e. manually restoring the state or just continue, since after this the only thing left is to call os::die/os::abort. If we reuse this logic with jfrEmergencyDump::on_vm_shutdown() then we would have to restore it, since I see print_bug_submit_message() which happens after that checks the state.